Default IItemHandler capability for shulker box itemstacks

lukebemish opened the pull request:
Forge provides an `IItemHandler` capability for most built-in block entities and entities, and provides fluid capabilities for buckets, but does not provide a built-in item capabillity for shulker box items. This PR includes a simple implementation of such a capability, as discussed a bit on discord; concerns about performance were taken into account mostly only when attaching the capability, as it does not seem likely that many shulker box items will be interacted with in this way in a single tick. Otherwise, the `IItemHandler` is very similar to the wrappers given to vanilla container inventories.

lukebemish committed:
Create wrapper class

lukebemish committed:
Rename wrapper

lukebemish committed:
Add capability to itemstacks

lukebemish committed:
Fix data being stored incorrectly in items

lukebemish committed:
Fix minor formatting details

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+import java.util.Set;
+
+@Mod.EventBusSubscriber(modid = "forge")
+public class ShulkerItemStackInvWrapper implements IItemHandler, ICapabilityProvider
```
Implementing `ICapabilityProvider` here makes this public API while it should be an implementation detail.
Additionally, `IItemHandlerModifiable` should be implemented as well.

diesieben07 replied:
Additionally, `IItemHandlerModifiable` should be implemented as well.

alcatrazEscapee replied:
Why should it be different than any of the other item stack capability wrappers that Forge providers? Considering that every item stack capability wrapper already follows this pattern, including [`FluidBucketWrapper`](https://github.com/MinecraftForge/MinecraftForge/blob/c2225ab1da58cf8eb3ac3aab1aab73f2d7a696fa/src/main/java/net/minecraftforge/fluids/capability/wrappers/FluidBucketWrapper.java#L32), [`FluidHandlerItemStack`](https://github.com/MinecraftForge/MinecraftForge/blob/c2225ab1da58cf8eb3ac3aab1aab73f2d7a696fa/src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerItemStack.java#L30), and [`FluidHandlerItemStackSimple`](https://github.com/MinecraftForge/MinecraftForge/blob/c2225ab1da58cf8eb3ac3aab1aab73f2d7a696fa/src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerItemStackSimple.java#L27).

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+        {
+            return getItemList().get(slot);
+        }
+        return ItemStack.EMPTY;
```
This should throw an exception instead.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    @NotNull
+    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate)
```
The logic here seems significantly more complex than the equivalent in `ItemStackHandler`. Why is that?

lukebemish replied:
This logic was copied from InvWrapper; it may be a sign that InvWrapper needs a refactor. I'll look at using the logic from ItemStackHandler instead.

lukebemish replied:
(Fixed by [4076409](https://github.com/MinecraftForge/MinecraftForge/pull/8827/commits/40764098c55531042d61406596c758b5a659664e))

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+        {
+            if (stackInSlot.getCount() < amount)
+            {
+                return stackInSlot.copy();
```
This stack was just deserialized from NBT, there is no need to copy it.

lukebemish replied:
(Fixed by [4076409](https://github.com/MinecraftForge/MinecraftForge/pull/8827/commits/40764098c55531042d61406596c758b5a659664e))

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+        return stack.getItem().canFitInsideContainerItems();
+    }
+
+    public NonNullList<ItemStack> getItemList()
+    {
+        NonNullList<ItemStack> itemStacks = NonNullList.withSize(getSlots(), ItemStack.EMPTY);
+        CompoundTag rootTag = BlockItem.getBlockEntityData(this.stack);
+        if (rootTag != null && rootTag.contains("Items", 9)) {
+            ContainerHelper.loadAllItems(rootTag, itemStacks);
+        }
+        return itemStacks;
+    }
+
+    public void setItemList(NonNullList<ItemStack> itemStacks)
+    {
+        CompoundTag existing = BlockItem.getBlockEntityData(this.stack);
+        BlockItem.setBlockEntityData(this.stack, BlockEntityType.SHULKER_BOX,
+                ContainerHelper.saveAllItems(existing == null ? new CompoundTag() : existing, itemStacks));
+    }
```
This should really use a cache and only deserialize when the stack's NBT tag has not changed.

lukebemish replied:
(Fixed by [8eb35a4](https://github.com/MinecraftForge/MinecraftForge/pull/8827/commits/8eb35a4984587249bfa5e8811c797466ba90b580), I think, though my caching setup could use a once-over)

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+            Items.YELLOW_SHULKER_BOX);
+
+    @SubscribeEvent
+    public static void listenCapabilitiesAttachment(AttachCapabilitiesEvent<ItemStack> event)
```
This should not be public API.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,247 @@
+        return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.orEmpty(cap, this.holder);
+    }
+
+    private static final Set<Item> SHULKER_ITEMS = Set.of(Items.SHULKER_BOX,
```
This sounds like a candidate for a tag.

lukebemish replied:
This was a discussion had on discord. While I agree that this could be a tag, we also discussed whether this should be an internal implementation; if it is an internal implementation, then it wouldn't make sense to make this a tag. (after all, other mods' shulker boxes might store items in nbt differently). I could see this going either way; just let me know what everyone thinks is best.

lukebemish replied:
(copied from discord) My personal opinion is that it should not, and the list of targets should be immutable, but that I could separate out the IItenHandler implementation from the ICapabilityProvider so that modders could implement the prior for their own items. (honestly, I think this is a bit ill-advised. My implementation is meant to work for vanilla items. Modded items should use something more efficient with a serializable cap or whatever)

lukebemish committed:
Change insert/extract implementation

lukebemish committed:
Use cache to decrease itemstack list deserialization

lukebemish committed:
Add slot index validation to getStackInSlot

lukebemish commented:
I went through and addressed most of your feedback; the one thing I am not sure how you'd like to have set up is the public API vs implementation details part. When I had written this, I'd intended it *all* to be basically implementation details; this wrapper is targeted at vanilla shulker boxes, and any modded items wouldn't be expected to use it. I can definitely separate out the ICapabilityProvider part and the listener somewhere else, so that that becomes an implementation detail and the capability something modders could use.

lukebemish commented:
I modeled the combined capability provider/item handler off of how Forge adds stuff to buckets at present, but I'm not sure what the best arrangement would be for that

lukebemish committed:
Delegate subscription to private class

Modeled after what was done in, i.e, ForgeSpawnEggItem

TheCurle commented:
Looks good to me for now.
It's currently the weekend, so we'll have a closer look tomorrow, and hopefully pull.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+    public ShulkerItemStackInvWrapper(ItemStack stack)
+    {
+        this.stack = stack;
+        this.itemStacksCache = refreshItemList(BlockItem.getBlockEntityData(this.stack));
```
Why load the data immediately here? It is not needed yet.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+
+import java.util.Set;
+
+public class ShulkerItemStackInvWrapper implements IItemHandlerModifiable, ICapabilityProvider
```
This class should not be public.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+    protected NonNullList<ItemStack> refreshItemList(CompoundTag rootTag)
+    {
+        NonNullList<ItemStack> itemStacks = NonNullList.withSize(getSlots(), ItemStack.EMPTY);
+        if (rootTag != null && rootTag.contains("Items", 9)) {
```
- Formatting, braces on new lines
- Use `Tag.TAG_LIST` instead of the magic number

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+    protected NonNullList<ItemStack> getItemList()
+    {
+        CompoundTag rootTag = BlockItem.getBlockEntityData(this.stack);
+        if ((cachedTag != null && !cachedTag.equals(rootTag)) || rootTag == null)
```
`Objects.equals`

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+    {
+        CompoundTag existing = BlockItem.getBlockEntityData(this.stack);
+        CompoundTag rootTag = ContainerHelper.saveAllItems(existing == null ? new CompoundTag() : existing, itemStacks);
+        BlockItem.setBlockEntityData(this.stack, BlockEntityType.SHULKER_BOX,
+                rootTag);
```
Why the line break?

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+                Items.YELLOW_SHULKER_BOX);
+
+        @SubscribeEvent
+        public static void listenCapabilitiesAttachment(AttachCapabilitiesEvent<ItemStack> event) {
```
This should use `initCapabilities` in the Item class, not the event.

lukebemish replied:
As discussed multiple times, shulker boxes use normal BlockItems.

lukebemish replied:
The only way to make this use initCapabilities would be to make an otherwise-unused ShulkerBoxBlockItem class and patch it in for shulkers, which seemed like overkill to me. If this is the preferred solution, please let me know

LexManos replied:
No, it's not. You *could* change  `public static void listenCapabilitiesAttachment(AttachCapabilitiesEvent<ItemStack> event) `  to `@ApiInternal @Nullable public static ICapabilityPrvider getVanillaShulkerInventory(Item) ` and call it from `IForgeItem.initCapabilities`'s default implementation.

Also, don't use a static set. Or make the set based on the delegate.

lukebemish replied:
As discussed on discord already, we *don't* want to use delegates here, because if someone registry-replaces the shulker boxes, we don't want to attach out capability.

LexManos replied:
And I disagree, because if you implement it in initCaps, then they can opt out via NOT calling  super.

lukebemish replied:
That doesn't solve the issue: imagine we attach this in `initCapabilities` checking equality with a delegate. Someone else replaces shulker boxes with their own normal `BlockItem`. We've now attached a capability to that BlockItem, even though they never asked us to.

LexManos replied:
That is their problem. Opt out vs Opt In. If they break vanilla API by replacing the item, and then breaking how they store data, that is on them.

lukebemish replied:
Alright, I should rephrase: if we do this with delegates, it becomes a breaking change. If you're alright with that, I'll go ahead and switch it to delegates then.

LexManos replied:
By that definition this is a breaking change by adding it at all.
So no, you're wrong. 
And i've already told you what to do. So do it. I've already told this to my triage team about 5 times.

diesieben07 commented on src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java:
```diff
@@ -0,0 +1,239 @@
+    @Mod.EventBusSubscriber(modid = "forge")
+    private static class AttachmentHandler
+    {
+        private static final Set<Item> SHULKER_ITEMS = Set.of(Items.SHULKER_BOX,
+                Items.BLACK_SHULKER_BOX,
+                Items.BLUE_SHULKER_BOX,
+                Items.BROWN_SHULKER_BOX,
+                Items.CYAN_SHULKER_BOX,
+                Items.GRAY_SHULKER_BOX,
+                Items.GREEN_SHULKER_BOX,
+                Items.LIGHT_BLUE_SHULKER_BOX,
+                Items.LIGHT_GRAY_SHULKER_BOX,
+                Items.LIME_SHULKER_BOX,
+                Items.MAGENTA_SHULKER_BOX,
+                Items.ORANGE_SHULKER_BOX,
+                Items.PINK_SHULKER_BOX,
+                Items.PURPLE_SHULKER_BOX,
+                Items.RED_SHULKER_BOX,
+                Items.WHITE_SHULKER_BOX,
+                Items.YELLOW_SHULKER_BOX);
```
The formatting here is weird, the items should all be on their own line

lukebemish committed:
Address formatting feedback

lukebemish committed:
Fix braces on newlines

lukebemish commented:
Except the whole `initCapabilities` thing, this should be good to go; if you want me to use `initCapabilities`, that would require creating an otherwise-unused `ShulkerBoxBlockItem` class and patching shulker boxes to use it. I could, but I'd like some confirmation first.

lukebemish committed:
Switch to initCapabilities

lukebemish commented:
Should be switched to `initCapabilities` now; I would note that this could be considered a breaking change now, though fairly minor. If a mod had previously registry-replaced shulker boxes with a version that stored less/more items, they would now need to use a custom BlockItem class that overrides the default behavior.

LexManos committed:
Update IForgeItem.java

LexManos committed:
Update ShulkerItemStackInvWrapper.java

LexManos reviewed (commented):
Change costly thing to after cheap thing

LexManos committed:
Change costly thing to happen after cheap ting
